---
title: "RNAseq with EdgeR"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
params:
  rmd: ""
output:
  html_document:
    dev: png
    code_folding: hide
    self_contained: yes
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: false
      smooth_scroll: true
    number_sections: true
    df_print: paged
---

```{r setup}
# log <- file(snakemake@log[[1]], open="wt")
# sink(log)
# sink(log, type="message")
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE,
	cache = TRUE
)
# for knit testing in RStudio server
#knitr::opts_knit$set(root.dir = "../")
```

```{r load_libs}
# my packages
library(here)
library(tidyverse)
library(vroom)
library(yaml)
library(kableExtra)
library(ggrepel)
# for bbcRNA tutorial
library(bbcRNA)
library(SummarizedExperiment)
# library(org.Mm.eg.db)# load the org.db for your organism
# library(org.Hs.eg.db)
library(enrichplot)
library(VennDiagram)
```

```{r config}
config <- yaml.load_file("config.yaml")
annotation <- read_tsv("annotation.tsv")
```

```{r}
# genes.nebraska <- vroom("analysis/star/nebraska_earlyBudding_102.ReadsPerGene.out.tab", 
#                          skip = 4, col_names = c("gene","unstranded","forward","reverse")) %>%
#   dplyr::select("gene")
# 
# genes.arikara <- vroom("analysis/star/arikara_earlyBudding_16.ReadsPerGene.out.tab", 
#                         skip = 4, col_names = c("gene","unstranded","forward","reverse")) %>%
#   dplyr::select("gene")
# 
# print(paste("unique to Arikara:",length(genes.arikara[["gene"]])))
# print(paste("unique to NE:",length(genes.nebraska[["gene"]])))
# 
# print(paste("unique to Arikara:",nrow(dplyr::anti_join(genes.arikara,genes.nebraska))))
# print(paste("unique to Nebraska:",nrow(dplyr::anti_join(genes.nebraska,genes.arikara))))
# print(paste("common to both:",nrow(dplyr::inner_join(genes.nebraska,genes.arikara))))

```

```{r import_counts}

if(config$strandedness == "unstranded"){
  star_col <- 1
}else if(config$strandedness == "forward"){
  star_col <- 2
}else if(config$strandedness == "reverse"){
  star_col <- 3
}else{stop("config$strandedness=='",config$strandedness,"'.This is an invalid value. Needs to be one of: 'unstranded','forward','reverse'")}

count_matrix.nebraska <- star_to_mat(dir = "../analysis/star/nebraska/",
                     rgx = "^[^_]+_[^_]+_[^\\.]+", column = star_col)
count_matrix.arikara <- star_to_mat(dir = "../analysis/star/arikara",
                     rgx = "^[^_]+_[^_]+_[^\\.]+", column = star_col)

#count_matrix <- count_matrix.nebraska
count_matrix <- dplyr::inner_join(as.data.frame(count_matrix.arikara) %>% rownames_to_column(var="gene"),
                  as.data.frame(count_matrix.nebraska) %>% rownames_to_column(var="gene")) %>%
  column_to_rownames(var="gene") %>%
  as.matrix()

```

# Experimental Design

```{r Make_BbcSE_object}
# bbc_obj <- BbcSE(counts = tcell, granges = granges)
bbc_obj <- BbcSE(counts = count_matrix)

# show more information about the BbcSE class
#getClassDef(class(bbc_obj))

# show object information
# bbc_obj

# get column metadata
#col_meta <- read_col_meta("../bin/units.tsv")
col_meta <- read.table(here(config$units),header=T) %>%
  dplyr::select(-c(fq1,fq2)) %>%
  dplyr::distinct() %>%
  remove_rownames() %>%
  column_to_rownames(var="sample")

# Add column meta data.
colData(bbc_obj) <- cbind(colData(bbc_obj), col_meta[colnames(bbc_obj),])

# view the meta data.
# colnames(colData(bbc_obj))
knitr::kable(colData(bbc_obj)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

# Quality Control

```{r get_mapping_rates}
aln_metrics <- read_star_aln_metrics(dir = "../analysis/star/",
                                 rgx = "^[^_]+_[^_]+_[^\\.]+")
# store the mapping metrics in the BbcSE object
aln_metrics(bbc_obj) <- aln_metrics
if (!validObject(bbc_obj)){
  knitr::knit_exit("ERROR: bbc_obj was no longer valid after adding aln_metrics")
}

kable(aln_metrics) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

```{r plot_alignment_rates, fig.width=12, fig.height=9}

# unique map reads
bbc_obj@aln_metrics
plot_aln_metrics(x = bbc_obj,
                 type = "uniq_aln_reads",
                 facet_by="tissue",
                 fill="accession") +
  ggplot2::theme(legend.position = "right")
#unique map rates
plot_aln_metrics(x = bbc_obj,
                 type = "uniq_map_rate",
                 facet_by="tissue",
                 fill="accession") +
  ggplot2::theme(legend.position = "right")

```

<!-- ```{r add_gene_annotations} -->

<!-- # Add gene symbols -->

<!-- bbc_obj <- ens2sym(bbc_obj, org.Mm.eg.db) -->

<!-- # Get Entrez IDs for gene set analyses after DE genes are identified -->

<!-- bbc_obj <- ens2entrez(bbc_obj, org.Mm.eg.db) -->

<!-- #rowData(bbc_obj) -->

<!-- if (!validObject(bbc_obj)){ -->

<!--   knitr::knit_exit("ERROR: bbc_obj was no longer valid after adding gene annotations") -->

<!-- } -->

<!-- ``` -->

```{r make_DGEList_object}
# by default, low expression genes filtered out, normalization factors
# calculated, normalized counts calculated
bbc_obj <- makeDGEList(bbc_obj, group="group")
# what's in the edger slot now?
# str(edger(bbc_obj))
# number of rows/features/genes in the SE
cat("*",nrow(bbc_obj),"total features/genes in the genome. \n\n")
# Number of genes in edger$DGEList
cat("*",nrow(dgelist(edger(bbc_obj))),"features/genes assessed after trimming.")
```

```{r PCA}
set.seed(100) # adonis uses permutations. Set seed to make reproducible.
pca_plots <- plot_PCA(bbc_obj, color_by="tissue",shape_by="accession", adonis_by="tissue")
pca_plots[[1]] #+
  # geom_label_repel(aes(label = replicate),
  #                  box.padding   = 0.35,
  #                  point.padding = 0.5,
  #                  segment.color = 'grey50')
pca_plots[[2]]

```

## Annotate Genes
```{r}
rowData(bbc_obj) <- data.frame(gene=rownames(bbc_obj)) %>% dplyr::mutate(gene=str_remove(gene,"gene:")) %>% DataFrame(.)

rowData(bbc_obj)
```

# Differential Expression

## Leaf

```{r}
#bbc_obj <- run_de(bbc_obj, contrast_name="leaf.Ari_over_NE")
bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_leaf","nebraska_leaf")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_leaf_-_nebraska_leaf`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("leaf", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-leaf.tsv")

```

## Meristem

```{r}
#bbc_obj <- run_de(bbc_obj, contrast_name="leaf.Ari_over_NE")
bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_meristem","nebraska_meristem")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_meristem_-_nebraska_meristem`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("meristem", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-meristem.tsv")

```

## Node

```{r}
#bbc_obj <- run_de(bbc_obj, contrast_name="leaf.Ari_over_NE")
bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_node","nebraska_node")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_node_-_nebraska_node`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-node.tsv")

```

## Root

```{r}

bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_root","nebraska_root")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_root_-_nebraska_root`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-root.tsv")

```

## Shoot

```{r}

bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_shoot","nebraska_shoot")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_shoot_-_nebraska_shoot`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-shoot.tsv")

```

## Seedcoat

```{r}

bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_seedcoat","nebraska_seedcoat")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_seedcoat_-_nebraska_seedcoat`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-seedcoat.tsv")

```

## Endosperm

```{r}
#bbc_obj <- run_de(bbc_obj, contrast_name="leaf.Ari_over_NE")
bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_endosperm","nebraska_endosperm")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_endosperm_-_nebraska_endosperm`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-endosperm.tsv")

```

## earlyBudding

```{r}

bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_earlyBudding","nebraska_earlyBudding")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_earlyBudding_-_nebraska_earlyBudding`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-earlyBudding.tsv")

```

## lateBudding

```{r}
#bbc_obj <- run_de(bbc_obj, contrast_name="leaf.Ari_over_NE")
bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = list(c("group","arikara_lateBudding","nebraska_lateBudding")),
                      lfc = config$logfc.filter)


# show pval distribution
print(plot_pval_distrib(bbc_obj))

# get results: gene, logFC, 
results.leaf <- cbind(as.data.frame(bbc_obj@edger@de_results$`group_arikara_lateBudding_-_nebraska_lateBudding`$table),
                      bbc_obj@edger@norm_cts@assays@data$norm_log_cpm %>% 
                        as.data.frame() %>% 
                        dplyr::select_if(grepl("node", names(.)))) %>%
  rownames_to_column(var="gene") %>% dplyr::mutate(gene= str_remove(gene,pattern="gene:")) %>%
  dplyr::select(-c(unshrunk.logFC))

# write results to file
write_tsv(results.leaf, path= "../deliverables/DE-lateBudding.tsv")

```





```{r , eval=FALSE, include=FALSE, results='asis'}
# import the contrasts.tsv
contrast_tbl = read_tsv(config$contrasts)

contrast_list <- list()
i=1
for (i in 1:nrow(contrast_tbl)){
  contrast_list[[i]] <- c(contrast_tbl[i,]$meta_col,
                          contrast_tbl[i,]$numerator,
                          contrast_tbl[i,]$denominator)
}

# loop through the comparisons
for (i in 1:length(contrast_tbl$name)){
  cat("# ",contrast_tbl$name[i],"\n\n")

  cat("## Differential Expression \n\n")

  cat("Differential Expression was performed using the `glmTreat` framework in `edgeR`. \n\n")
  cat("First, we check the distribution of p-values from the performed tests to confirm that the tests ran successfully. For `glmTreat` tests, we expect to see a steep slope downward from p=0 and a rise toward p=1. The rise for very high p-values represents the genes with differential expression values less than our defined cutoff value, log2FC >", config$logfc.filter,".\n\n")

  # run the comparison
  bbc_obj <- findDEGs(bbc_obj,
                      de_pkg = "edger",
                      test = "glmTreat",
                      design = "~0+group",
                      contrasts = contrast_list[i],
                      lfc = config$logfc.filter)

    # plot p-val distribution -------------------------------------------------------------
    print(plot_pval_distrib(bbc_obj))
    #plot_volcano(bbc_obj)  
    cat("\n\n")
# 
    # get DE table------------------------------------------------------------------------
    de_res <- get_de_table(bbc_obj, de_pkg = "edger")

    # write the DE table to file
    write_delim(de_res[[1]],
                paste0("deliverables/DE_",contrast_tbl$name[i],".tsv"),
                delim="\t")
    # print 20 top_genes from DE table
    top_genes <- edgeR::topTags(
      de_results(edger(bbc_obj))[[2]],
      n=20)



    print(kable(de_res[[1]] %>%
                  dplyr::filter(ensembl_id %in% rownames(top_genes$table)),
                caption="Top 20 Differentially Expressed Genes") %>%
      kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")))
# 
#     # plot heatmap-------------------------------------------------------------------------
# 
#     cat("### Heatmap \n\n")
# 
#     cat("Clustered Heatmap of Top 20 Differentially Expressed Genes. \n\n")
#     # # get the sign of the LFC and store in rowData
#     # ## get the de_results slot of the BbcEdgeR object
#     # bbc_de_res <- de_results(edger(bbc_obj))
#     # ## get contrast results
#     # de_table <- bbc_de_res[[2]]$table
#     # ## get the LFC sign
#     # contrast_sign <- ifelse(de_table$logFC > 0, "up", "down")
#     # names(contrast_sign) <- rownames(de_table)
#     # ## store LFC sign in rowData. Genes with no matches will be NA.
#     # rowData(bbc_obj)$mut_vs_WT_sign <- contrast_sign[match(rownames(bbc_obj), names(contrast_sign))]
#     # paste0(contrast_tbl[i,]$relative,"_vs_",contrast_tbl[i,]$baseline,"_sign")
#     # plot Zscores of normalized log2 cpms
#     print(plot_heatmap(x = bbc_obj, genes = rownames(top_genes$table), zscores=TRUE,
#                  # rowdata_annot = paste0(contrast_tbl[i,]$relative,"_vs_",
#                  #                        contrast_tbl[i,]$baseline,"_sign"),
#                  # coldata_annot = "group",
#                  # rowdata_split = paste0(contrast_tbl[i,]$relative,"_vs_",
#                  #                        contrast_tbl[i,]$baseline,"_sign"),
#                  # coldata_split="",
#                  gene_labels = "uniq_syms"))
#     cat("\n\n")
#     # # Enrichment Analysis-------------------------------------------------------------------------
#     # cat("### Enrichment Analysis \n\n")
#     # # set seed to make GSEA deterministic
#     # set.seed(42)
#     # 
#     # # Hallmarks
#     # H_gsea_results_list <- run_gsea(x = bbc_obj,
#     #                                 de_pkg = "edger",
#     #                                 gene_set = "H",
#     #                                 orgDb = dplyr::filter(annotation, common_name==config$common_name)$org.db,
#     #                                 organism = dplyr::filter(annotation, common_name==config$common_name)$species)
#     # 
#     # if (length(H_gsea_results_list[[1]]@result$ID) > 0){
#     #   enrichplot::dotplot(H_gsea_results_list[[1]],
#     #                       showCategory=10,
#     #                       title="Top gene sets",
#     #                       split=".sign") +
#     #     facet_grid(.~.sign)
#     # } else{cat("No significant enrichment of Hallmark Gene Sets. \n\n")}
#     # 
#     # # KEGG
#     # kegg_gsea_results_list <- run_gsea(x = bbc_obj,
#     #                             de_pkg = "edger",
#     #                             gene_set = "kegg",
#     #                             orgDb = dplyr::filter(annotation, common_name==config$common_name)$org.db,
#     #                             organism = dplyr::filter(annotation, common_name==config$common_name)$kegg,
#     #                             minGSSize = 30,
#     #                             use_internal_data = FALSE)
#     # 
#     #   if (length(kegg_gsea_results_list[[1]]@result$ID) > 0){
#     #     enrichplot::dotplot(kegg_gsea_results_list[[1]],
#     #                         showCategory=10,
#     #                         title="Top gene sets",
#     #                         split=".sign") +
#     #       facet_grid(.~.sign)
#     # } else{cat("No significant enrichment of KEGG Gene Sets. \n\n")}
}
```

# Session Info

```{r sessionInfo}
sessionInfo()
```
