---
title: "RNAseq with EdgeR"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
params:
  rmd: ""
output:
  html_document:
    dev: png
    code_folding: hide
    self_contained: yes
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: false
      smooth_scroll: true
    number_sections: true
    df_print: paged
---

```{r setup}
# log <- file(snakemake@log[[1]], open="wt")
# sink(log)
# sink(log, type="message")
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE,
	cache = TRUE
)
# for knit testing in RStudio server
#knitr::opts_knit$set(root.dir = "../")
```

```{r load_libs}
# my packages
library(here)
library(tidyverse)
library(vroom)
library(yaml)
library(kableExtra)
library(ggrepel)
# for bbcRNA tutorial
library(bbcRNA)
library(SummarizedExperiment)
# library(org.Mm.eg.db)# load the org.db for your organism
# library(org.Hs.eg.db)
library(enrichplot)
```

```{r config}
config <- yaml.load_file("config.yaml")
annotation <- read_tsv("annotation.tsv")
```

```{r functions}
get_DE_results <- function(summarizedExperiment, tissue, contrast.group1, contrast.group2, logFC){
  # setup contrast name from experimental groups (contrast.group1 is numerator, contrast.group2 is denominator)
  contrast.name = paste0("group_",contrast.group1,"_",tissue,"_-_",contrast.group2,"_",tissue)
  
  # run DE with bbcRNA function findDEGs() using glmQLFTest function and logFC cutoff from config (default logFC==1)
  summarizedExperiment.wDE = findDEGs(summarizedExperiment,
                                     de_pkg = "edger",
                                     test = "glmQLFTest",
                                     design = "~0+group",
                                     contrasts = list(c("group", paste0(contrast.group1,"_",tissue), paste0(contrast.group2,"_",tissue))),
                                     lfc = logFC)

  # get the contents from the de_results slot, removing the fit object (1st element)
  DE_table = de_results(edger(summarizedExperiment.wDE))[-1]
  
  # make results data.frame with DE stats and logCPM values
  DE_results = dplyr::left_join(
    # get results for current contrast.name
    edgeR::topTags(DE_table[[contrast.name]],
                   # include all genes in results
                   n = nrow(dgelist(edger(summarizedExperiment.wDE)))) %>%
      as.data.frame(stringsAsFactors = FALSE) %>% 
      rownames_to_column(var="gene"),
    # and join logCPM values 
    summarizedExperiment.wDE@edger@norm_cts@assays@data$norm_log_cpm %>% 
      as.data.frame() %>% 
      select_if(grepl(tissue, names(.))) %>% 
      rownames_to_column(var="gene"),
    # join by "gene" column
    by="gene") %>%
  # and remove prefix "gene:" from gene column
  dplyr::mutate(gene=str_remove(gene,"gene:"))
  
  # return both summarizedExperiment object and DE_results data.frame in a list
  return(list(summarizedExperiment.wDE, DE_results))
}

plot.volcano <-function(DE_results.tissue){
  DE_results.tissue[[2]] %>% 
    # add significance and logFC criteria for plotting
    dplyr::mutate(Significance = factor(case_when(
      (FDR < 0.05) & (abs(logFC) > 1) ~ "logFC>1 & FDR<0.05",
      (FDR > 0.05) & (abs(logFC) > 1) ~ "logFC>1 only",
      (FDR < 0.05) & (abs(logFC) < 1) ~ "FDR<0.05 only",
      (FDR > 0.05) & (abs(logFC) < 1) ~ "NS"),
      levels = c("logFC>1 & FDR<0.05","logFC>1 only","FDR<0.05 only", "NS"))) %>%
    # now ggplot
    ggplot(.) +
      geom_point(aes(x=logFC, y=-log10(FDR), color=Significance, label=gene)) +
      theme_light() + ggthemes::scale_color_wsj(palette = "rgby")
}

plot.MD <-function(DE_results.tissue){
  DE_results.tissue[[2]] %>% 
    # add significance and logFC criteria for plotting
    dplyr::mutate(Significance = factor(case_when(
      (FDR < 0.05) & (abs(logFC) > 1) ~ "logFC>1 & FDR<0.05",
      (FDR > 0.05) & (abs(logFC) > 1) ~ "logFC>1 only",
      (FDR < 0.05) & (abs(logFC) < 1) ~ "FDR<0.05 only",
      (FDR > 0.05) & (abs(logFC) < 1) ~ "NS"),
      levels = c("logFC>1 & FDR<0.05","logFC>1 only","FDR<0.05 only", "NS"))) %>%
    # now ggplot
    ggplot(.) +
      geom_point(aes(x=logCPM, y=logFC, color=Significance, label=gene)) +
      theme_light() + ggthemes::scale_color_wsj(palette = "rgby")
}
```

## Compare References

```{r}
genes.HanXRQr2 <- read_tsv("../analysis/ref-gene-compare/HanXRQr2.0-SUNRISE-2.1.genes.tsv",col_names = "gene") %>% 
  dplyr::mutate(reference="HanXRQr2")
genes.arikara <- read_tsv("../analysis/ref-gene-compare/arikara.genes.tsv",col_names = "gene") %>% 
  dplyr::mutate(gene=str_remove(gene, pattern='";')) %>% 
  dplyr::mutate(reference="arikara")
genes.nebraska <- read_tsv("../analysis/ref-gene-compare/nebraska.genes.tsv",col_names = "gene") %>% 
  dplyr::mutate(gene=str_remove(gene, pattern='";'))%>% 
  dplyr::mutate(reference="nebraska")

nrow(genes.HanXRQr2)
nrow(genes.arikara)
nrow(genes.nebraska)

nrow(dplyr::inner_join(genes.nebraska,genes.arikara, by="gene"))

# genes unique to nebraska
paste("Unique to Nebraska:",
      nrow(dplyr::anti_join(genes.nebraska,genes.arikara, by="gene")))
write_tsv(dplyr::anti_join(genes.nebraska,genes.arikara, by="gene"),
          path= "../deliverables/unique-genes.nebraska.tsv")
# genes unique to arikara
paste("Unique to Arikara:",
      nrow(dplyr::anti_join(genes.arikara,genes.nebraska, by="gene")))
write_tsv(dplyr::anti_join(genes.arikara,genes.nebraska, by="gene"),
          path= "../deliverables/unique-genes.arikara.tsv")
```

```{r import_counts}

if(config$strandedness == "unstranded"){
  star_col <- 1
}else if(config$strandedness == "forward"){
  star_col <- 2
}else if(config$strandedness == "reverse"){
  star_col <- 3
}else{stop("config$strandedness=='",config$strandedness,"'.This is an invalid value. Needs to be one of: 'unstranded','forward','reverse'")}

count_matrix.nebraska <- star_to_mat(dir = "../analysis/star/nebraska/",
                     rgx = "^[^_]+_[^_]+_[^\\.]+", column = star_col)
count_matrix.arikara <- star_to_mat(dir = "../analysis/star/arikara",
                     rgx = "^[^_]+_[^_]+_[^\\.]+", column = star_col)

#count_matrix <- count_matrix.nebraska
count_matrix <- dplyr::inner_join(as.data.frame(count_matrix.arikara) %>% rownames_to_column(var="gene"),
                  as.data.frame(count_matrix.nebraska) %>% rownames_to_column(var="gene")) %>%
  column_to_rownames(var="gene") %>%
  as.matrix()

```

# Experimental Design

```{r Make_BbcSE_object}
# bbc_obj <- BbcSE(counts = tcell, granges = granges)
bbc_obj <- BbcSE(counts = count_matrix)

# show more information about the BbcSE class
#getClassDef(class(bbc_obj))

# show object information
# bbc_obj

# get column metadata
#col_meta <- read_col_meta("../bin/units.tsv")
col_meta <- read.table(here(config$units),header=T) %>%
  dplyr::select(-c(fq1,fq2)) %>%
  dplyr::distinct() %>%
  remove_rownames() %>%
  column_to_rownames(var="sample")

# Add column meta data.
colData(bbc_obj) <- cbind(colData(bbc_obj), col_meta[colnames(bbc_obj),])

# view the meta data.
# colnames(colData(bbc_obj))
knitr::kable(colData(bbc_obj)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

# Quality Control

```{r get_mapping_rates}
aln_metrics <- read_star_aln_metrics(dir = "../analysis/star/",
                                 rgx = "^[^_]+_[^_]+_[^\\.]+")
# store the mapping metrics in the BbcSE object
aln_metrics(bbc_obj) <- aln_metrics
if (!validObject(bbc_obj)){
  knitr::knit_exit("ERROR: bbc_obj was no longer valid after adding aln_metrics")
}

kable(aln_metrics) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

```{r plot_alignment_rates, fig.width=12, fig.height=9}

# unique map reads
bbc_obj@aln_metrics
plot_aln_metrics(x = bbc_obj,
                 type = "uniq_aln_reads",
                 facet_by="tissue",
                 fill="accession") +
  ggplot2::theme(legend.position = "right")
#unique map rates
plot_aln_metrics(x = bbc_obj,
                 type = "uniq_map_rate",
                 facet_by="tissue",
                 fill="accession") +
  ggplot2::theme(legend.position = "right")

```

<!-- ```{r add_gene_annotations} -->

<!-- # Add gene symbols -->

<!-- bbc_obj <- ens2sym(bbc_obj, org.Mm.eg.db) -->

<!-- # Get Entrez IDs for gene set analyses after DE genes are identified -->

<!-- bbc_obj <- ens2entrez(bbc_obj, org.Mm.eg.db) -->

<!-- #rowData(bbc_obj) -->

<!-- if (!validObject(bbc_obj)){ -->

<!--   knitr::knit_exit("ERROR: bbc_obj was no longer valid after adding gene annotations") -->

<!-- } -->

<!-- ``` -->

```{r make_DGEList_object}
# by default, low expression genes filtered out, normalization factors
# calculated, normalized counts calculated
bbc_obj <- makeDGEList(bbc_obj, group="group")
# what's in the edger slot now?
# str(edger(bbc_obj))
# number of rows/features/genes in the SE
cat("*",nrow(bbc_obj),"total features/genes in the genome. \n\n")
# Number of genes in edger$DGEList
cat("*",nrow(dgelist(edger(bbc_obj))),"features/genes assessed after trimming.")
```

## PCA

```{r PCA}
set.seed(100) # adonis uses permutations. Set seed to make reproducible.
pca_plots <- plot_PCA(bbc_obj, color_by="tissue",shape_by="accession", adonis_by="tissue")
pca_plots[[1]] #+
  # geom_label_repel(aes(label = replicate),
  #                  box.padding   = 0.35,
  #                  point.padding = 0.5,
  #                  segment.color = 'grey50')
pca_plots[[2]]

```

```{r}
# get the loadings
norm_counts <- norm_cts(edger(bbc_obj))
pca.loadings <- prcomp(t(norm_counts@assays@data$norm_log_cpm))
pca.loadings$rotation
```


```{r annotate_genes}
rowData(bbc_obj) <- data.frame(gene=rownames(bbc_obj)) %>% dplyr::mutate(gene=str_remove(gene,"gene:")) %>% DataFrame(.)
```


# Differential Expression

## QC

```{r}
DE_results.test <- get_DE_results(bbc_obj, "leaf", "arikara", "nebraska", config$logfc.filter)
```

```{r}
## Biological Coefficient of Variation
paste("Biological Coefficient of Variation:", sqrt(DE_results.test[[1]]@edger@dgelist$common.dispersion))
edgeR::plotBCV(DE_results.test[[1]]@edger@dgelist)

```

## Leaf

```{r}
# get DE table
DE_results.leaf <- get_DE_results(bbc_obj, "leaf", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.leaf[[2]], path= "../deliverables/DE-leaf.tsv")
```

```{r}
plot_pval_distrib(DE_results.leaf[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.leaf)
# MD plot
plot.MD(DE_results.leaf)
```

## Meristem

```{r}
# get DE table
DE_results.meristem <- get_DE_results(bbc_obj, "meristem", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.meristem[[2]], path= "../deliverables/DE-meristem.tsv")
```

```{r}
plot_pval_distrib(DE_results.meristem[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.meristem)
# MD plot
plot.MD(DE_results.meristem)
```

## Node

```{r}
# get DE table
DE_results.node <- get_DE_results(bbc_obj, "node", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.node[[2]], path= "../deliverables/DE-node.tsv")
```

```{r}
plot_pval_distrib(DE_results.node[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.node)
# MD plot
plot.MD(DE_results.node)
```

## Root

```{r}
# get DE table
DE_results.root <- get_DE_results(bbc_obj, "root", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.root[[2]], path= "../deliverables/DE-root.tsv")
```

```{r}
plot_pval_distrib(DE_results.root[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.root)
# MD plot
plot.MD(DE_results.root)
```

## Shoot

```{r}
# get DE table
DE_results.shoot <- get_DE_results(bbc_obj, "shoot", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.shoot[[2]], path= "../deliverables/DE-shoot.tsv")
```

```{r}
plot_pval_distrib(DE_results.shoot[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.shoot)
# MD plot
plot.MD(DE_results.shoot)
```

## Seedcoat

```{r}
# get DE table
DE_results.seedcoat <- get_DE_results(bbc_obj, "seedcoat", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.seedcoat[[2]], path= "../deliverables/DE-seedcoat.tsv")
```

```{r}
plot_pval_distrib(DE_results.seedcoat[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.seedcoat)
# MD plot
plot.MD(DE_results.seedcoat)
```

## Endosperm

```{r}
# get DE table
DE_results.endosperm <- get_DE_results(bbc_obj, "endosperm", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.endosperm[[2]], path= "../deliverables/DE-endosperm.tsv")
```

```{r}
plot_pval_distrib(DE_results.endosperm[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.endosperm)
# MD plot
plot.MD(DE_results.endosperm)
```

## earlyBudding

```{r}
# get DE table
DE_results.earlyBudding <- get_DE_results(bbc_obj, "earlyBudding", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.earlyBudding[[2]], path= "../deliverables/DE-earlyBudding.tsv")
```

```{r}
plot_pval_distrib(DE_results.earlyBudding[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.earlyBudding)
# MD plot
plot.MD(DE_results.earlyBudding)
```

## lateBudding

```{r}
# get DE table
DE_results.lateBudding <- get_DE_results(bbc_obj, "lateBudding", "arikara", "nebraska", config$logfc.filter)
# write results data.frame to deliverables/
write_tsv(DE_results.lateBudding[[2]], path= "../deliverables/DE-lateBudding.tsv")
```

```{r}
plot_pval_distrib(DE_results.lateBudding[[1]],de_method = "edger")
# volcano plot
plot.volcano(DE_results.lateBudding)
# MD plot
plot.MD(DE_results.lateBudding)
```

## RData
```{r}
write_tsv(as.data.frame(assay(bbc_obj@edger@norm_cts)) %>% rownames_to_column(var="gene"), "../deliverables/norm_log_cpm.tsv")
save.image(file = "diffExp.RData")
```


## SessionInfo()

```{r sessionInfo}
sessionInfo()
```
